### ISSUE TITLE
[FR] MLflow Tracing for Haystack

### ISSUE THREAD

**Issue by B-Step62** [mlflow/mlflow#13910]
### Summary

Expand [auto-tracing integration](https://mlflow.org/docs/latest/llms/tracing/index.html#automatic-tracing) of MLflow Tracing to [Haystack](https://haystack.deepset.ai/). 

### Required Changes
Please refer to [how to add new integration to MLflow Tracing](https://mlflow.org/docs/latest/llms/tracing/contribute.html) for the actual steps of adding new auto-tracing integration. MLflow maintainers will provide attentive support for designing and implementing the change.

### Expected Behavior

1. Users can enable auto-tracing by calling `mlflow.haystack.autolog()`
```python
import mlflow

from haystack import Pipeline
from haystack.components.generators import OpenAIGenerator
from haystack.components.builders.prompt_builder import PromptBuilder

mlflow.haystack.autolog()

# Initialize the pipeline
pipeline = Pipeline()

llm = OpenAIGenerator(model="gpt-4o-mini")
prompt_template = """Answer the question. {{question}}"""
prompt_builder = PromptBuilder(template=prompt_template)

pipeline = Pipeline()
pipeline.add_component("prompt_builder", prompt_builder)
pipeline.add_component("llm", llm)
pipeline.connect("prompt_builder", "llm")

question = "Who lives in Paris?"
results = pipeline.run({"prompt_builder": {"question": question}})
```
2. This will generates a trace like this (This example is taken from DSpy tracing. The structure will be different for Haystack tracing):
![image](https://github.com/user-attachments/assets/c6c636fe-e90e-4f0e-ada3-151717c334a7)

### Contact

If you have any question or willing to work on this feature, leave a comment with tagging @B-Step62

**Comment by SaumilPatel03** [comment]
@B-Step62 I’m new to the community, and I’m willing to work on this feature request.

**Comment by preyasshah9** [comment]
Hi @B-Step62 I'm new to `mlflow`. I'm willing to work on this feature, let me know if that sounds good to you!

**Comment by B-Step62** [comment]
@preyasshah9 Absolutely! Thank you for the willingness to contribute. Please let us know when you cave out the high level approach to implement this integration!

**Comment by aimlnerd** [comment]
Looking forward to Haystack support for MLflow tracing. Many ml platforms already have mlflow inbuilt, so this makes it enabling tracing for llm tracing "free" and "automatic". Instead of other paid tracing solutions.

**Comment by B-Step62** [comment]
Hi @preyasshah9! Did you have a chance to look into this initiative? Happy to discuss further if you have some high-level design and provide initial guidance about MLflow code base!😄

**Comment by preyasshah9** [comment]
Hi @B-Step62
I did go through the documentation on adding your own integration. I have some high level thoughts and would love to get some guidance on this from you! Since haystack is an orchestration (langchain, llamaindex), it definitely is more involved than I originally thought it would be. 
Let me share with you some initial thoughts on design by tomorrow.

**Comment by preyasshah9** [comment]
Hi @B-Step62 
There are a few open questions I have. 

- If Haystack has the [telemetry](https://docs.haystack.deepset.ai/docs/telemetry) support, does that serve as the equivalent of the callback functions that are required for implementing client APIs?
- In Design doc template, there is a REST API (full support) v/s Client based decision matrix. I do not have full context on the REST API based approach. Could you provide some guidelines on how to fill it out?

**Comment by preyasshah9** [comment]
Hi @B-Step62. I would like to get some initial guidance on MLFlow code base. Let me know what would be the best way to get some guidance from you.

**Comment by B-Step62** [comment]
@preyasshah9 Sorry for getting back late.

> If Haystack has the [telemetry](https://docs.haystack.deepset.ai/docs/telemetry) support, does that serve as the equivalent of the callback functions that are required for implementing client APIs? Or does this require defining your own custom backend as per specified in documentation: https://docs.haystack.deepset.ai/docs/tracing#custom-tracing-backend

The former telemetry page linked is not about general tracing, but their own telemetry for usage data (i.e. they collect how we use Haystack).

The latter seems to be promising. I think we can define a custom classes that inherits `Span` and `Tracer` base classes.

> In Design doc template, there is a REST API (full support) v/s Client based decision matrix. I do not have full context on the REST API based approach. Could you provide some guidelines on how to fill it out?

Please feel free to ignore those sections, they are not relevant to the tracing design.

> I would like to get some initial guidance on MLFlow code base. Let me know what would be the best way to get some guidance from you.

Sure, are you in [MLflow slack](https://mlflow.org/community/#slack) workspace? If so, we can continue on DM or huddle🙂

**Comment by preyasshah9** [comment]
I'm not in the MLFlow slack (Access Denied).

**Comment by B-Step62** [comment]
@preyasshah9 Sorry I pasted the wrong link. Could you check https://go.mlflow.org/slack instead? It is public channel.

**Comment by preyasshah9** [comment]
> [@preyasshah9](https://github.com/preyasshah9) Sorry I pasted the wrong link. Could you check https://go.mlflow.org/slack instead? It is public channel.

Actually the link is no longer valid :( Could you resend the link?

**Comment by B-Step62** [comment]
@preyasshah9 Oops sorry, what about this link? https://mlflow.org/slack

**Comment by y-okt** [comment]
@B-Step62 Hi, if no one is actively working on this, may I start working on this issue?

**Comment by bilgeyucel** [comment]
@B-Step62 Any updates on this one? I'm happy to support on the Haystack side and answer any questions! 🙌

**Comment by B-Step62** [comment]
@y-okt Thank you for offering the contribution! Yes, the previous assignee doesn't seem to be actively working on it, so it would be great if you could take it over.

I think you have good context for the tracing integration as you've implemented another one, but feel free to reach out for any questions!

@bilgeyucel Yes, your support would be much appreciated. Let us tag you on questions and technical reviews, thank you so much for the support!

**Comment by y-okt** [comment]
Thank you very much @B-Step62 and @bilgeyucel! I'll work on this and if any questions arises, let me ask. Thanks!
